Subscribe to price depth updates

nevmerzhitsky opened the pull request:
Первая версия фреймворка. Без тестов. Их добавлю позже или в другом ПР. Решил создать этот, чтобы ты мог дать обратную связь по коду, пока я делаю тесты.

Самый сложный код в `transport/`. Понять, как взаимодействуют классы, думаю, проще через анализ кода в `examples/` и далее вглубь.

nevmerzhitsky committed:
Init commit

nevmerzhitsky committed:
Fix license and add a file for docker build

nevmerzhitsky committed:
Fix minimal Python version

nevmerzhitsky committed:
Little improve of README

nevmerzhitsky commented on .dockerignore:
```diff
@@ -0,0 +1,4 @@
+.git/
```
Test

seriousidea commented on src/galts_trade_api/exchange.py:
```diff
@@ -0,0 +1,97 @@
+    def tag(self):
+        return self._tag
+
+    def add_market(self, market: Market):
```
Не хватает тайп-хинта.

seriousidea replied:
Не хватает тх.

nevmerzhitsky committed:
Refactor dockerfile to reduce cache invalidation chance

nevmerzhitsky committed:
Refactor done-callback for a router task in the main process

Now it fall to an exception handler instead of delivered on_exception()
callback. Also it calls shutdown before propogate an exception.

nevmerzhitsky committed:
Totally replace manual callback by specified loop exception handler

Also added handling of a signal from OS to gracefully shutdown
a program. The next step is make the technical code be a part of
the framework.

nevmerzhitsky committed:
Separate signal handling and shutdown tasks

nevmerzhitsky committed:
Extract technical code to a new package

seriousidea commented on src/galts_trade_api/asyncio_helper.py:
```diff
@@ -0,0 +1,85 @@
+
+class AsyncProgramEnv:
+    """
+    Main purpose of the class is give an ability to "patch" an event loop exception handler.
```
Кажется, не хватает `to` перед `give`.

seriousidea commented on src/galts_trade_api/asyncio_helper.py:
```diff
@@ -0,0 +1,85 @@
+LoopExceptionHandlerCallable = Callable[[asyncio.AbstractEventLoop, Dict[str, Any]], Any]
+
+
+async def signal_handler(signal, loop: asyncio.AbstractEventLoop):
```
Не хватает пары тайп-хинтов (тх).

seriousidea commented on src/galts_trade_api/asyncio_helper.py:
```diff
@@ -0,0 +1,85 @@
+
+
+# Inspired by https://www.roguelynn.com/words/asyncio-exception-handling/
+async def shutdown(loop: asyncio.AbstractEventLoop):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/asyncio_helper.py:
```diff
@@ -0,0 +1,85 @@
+# Inspired by https://www.roguelynn.com/words/asyncio-exception-handling/
+async def shutdown(loop: asyncio.AbstractEventLoop):
+    """Cleanup tasks tied to the program's shutdown."""
+    print('Shutting down...')
```
Про принты я пока ничего не буду писать. Понимаю что в следующих коммитах ты их удалишь.

nevmerzhitsky replied:
Да, планирую на логгер заменить.

seriousidea commented on src/galts_trade_api/asyncio_helper.py:
```diff
@@ -0,0 +1,85 @@
+
+        self._exception_handler_patch = value
+
+    def exception_handler(self, loop: asyncio.AbstractEventLoop, context: Dict[str, Any]):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/terminal.py:
```diff
@@ -0,0 +1,93 @@
+
+    async def _on_init_exchange_entities_response(
+        self,
+        event: MutableMapping[str, MutableMapping]
```
А почему `event`, если это не столько событие, сколько данные? Может лучше что-то типа `event_data`?

seriousidea commented on src/galts_trade_api/exchange.py:
```diff
@@ -0,0 +1,97 @@
+    def __init__(
+        self,
+        transport_factory: TransportFactory,
+        id: int,
```
Думаю что про предупреждения IDE "shadows built-in name" упоминать не буду (кроме случаев если это может быть не безопасно). Но интересно понять почему здесь ты решил использовать `id` и проигнорировать предупреждение, а в методе `Terminal._on_init_exchange_entities_response` ты используешь `id_`?

nevmerzhitsky replied:
Это случайность, что сделал по разному в двух местах. Хоть IDE и раздражает этим предупреждением в случае `id`, лично мне хочется использовать такое имя для переменной. Хотя с другой стороны, хочется видеть 0 предупреждений в IDE. И второе, кажется, побеждает, так что заменю на `id_`.

А ты сам как считаешь?

nevmerzhitsky replied:
А нет, я вспомнил. Здесь используется `id`, чтобы можно было вызывать конструктор в нотации `bla(**data)`. В `data` получается из клиента gRPC и я не хочу заменять "id" на "id_" в этом словаре перед инстанциированием. Так что я предпочту оставить код как сейчас.

seriousidea commented on src/galts_trade_api/tools.py:
```diff
@@ -0,0 +1,32 @@
+        return cls._instances[key]
+
+    @classmethod
+    def _calculate_args_hash(mcs, args, kwargs):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/transport/__init__.py:
```diff
@@ -0,0 +1,111 @@
+
+
+class DepthConsumeKey(NamedTuple):
+    exchange: str = '*'
```
А почему не `exchange_tag`?

seriousidea commented on src/galts_trade_api/transport/__init__.py:
```diff
@@ -0,0 +1,111 @@
+            request, data = message
+            self._notify(request, data)
+
+    def _notify(self, request: PipeRequest, data: Any) -> None:
```
С точки зрения самодокументируемого кода кажется что "_уведомить" - это не очень подходящее название по двум причинам:
1. Не понятно кого и о чем уведомить
2. Внутри метода происходит не уведомление потребителя, а передача ему данных

Может лучше что-то типа `pass_data_to_consumer`/`pass_response_to_consumer`?

nevmerzhitsky replied:
Согласен, что "уведомлять" тут не особо подходит. Переименовал переменную, этот метод, и метод в `MessageConsumerCollection`, чтобы исправить ситуацию.

Только pass здесь не подходит по смыслу. Словари говорят, что у него значение "пройти мимо", "пройти вдоль", "пройти в направлении" или "пройти сквозь". Я использовал в классе роутера "dispatch" (это типовой термин для кода про роутинг) и "send" в MessageConsumerCollection.

seriousidea commented on src/galts_trade_api/transport/__init__.py:
```diff
@@ -0,0 +1,111 @@
+
+        asyncio.create_task(self._consumers[request].notify(data))
+
+    def init_response_consumer(self, request: PipeRequest) -> MessageConsumerCollection:
```
Не могу перевести название этого метода чтобы оценить семантику. Что именно здесь инициализируется?
- Ответ потребителя? (`init_consumer_response`)
- Ответ потребителю? (`init_response_to_consumer`)
- Коллекция потребителей?  (`init_consumer_collection_for_response`)
- etc

nevmerzhitsky replied:
Я имел в виду "проинициализировать получателя ответа". Подумаю позже о переименовании.

nevmerzhitsky replied:
Переименовал в prepare_consumers_of_response

seriousidea commented on src/galts_trade_api/terminal.py:
```diff
@@ -0,0 +1,93 @@
+
+from .asset import Asset, Symbol
+from .exchange import Exchange, Market
+from .transport import OnExceptionCallable, TransportFactory
```
`OnExceptionCallable` и `Optional` не используются на данный момент.

seriousidea commented on src/galts_trade_api/transport/real.py:
```diff
@@ -0,0 +1,297 @@
+        exchange_info_get_entities_timeout: Optional[float] = None,
+        depth_scraping_queue_dsn: Optional[str] = None,
+        depth_scraping_queue_exchange: Optional[str] = None,
+    ):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/transport/real.py:
```diff
@@ -0,0 +1,297 @@
+        self._response_router = PipeResponseRouter(self._parent_connection)
+        task = asyncio.create_task(self._response_router.start())
+
+        def task_done_cb(t: asyncio.Task):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/transport/real.py:
```diff
@@ -0,0 +1,297 @@
+    async def _main(self) -> None:
+        self._ready_event.set()
+
+        async def loop():
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/transport/real.py:
```diff
@@ -0,0 +1,297 @@
+        async def loop():
+            stopped = False
+
+            def task_done_cb(t: asyncio.Task):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/transport/real.py:
```diff
@@ -0,0 +1,297 @@
+    def _notify_about_exception(self, e: Exception) -> None:
+        self._connection.send([e])
+
+    async def init_exchange_entities(self, request: InitExchangeEntitiesRequest):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/transport/real.py:
```diff
@@ -0,0 +1,297 @@
+        return result
+
+
+class RealTransportFactoryProcess(Process):
```
В этом классе наверное нужно изменить порядок методов, т.к. ранее мы договаривались о том, что все приватные методы должны идти после публичных.

seriousidea commented on examples/example.py:
```diff
@@ -0,0 +1,73 @@
+from galts_trade_api.transport.real import RealTransportFactory
+
+
+async def start_trade_system(program_env: AsyncProgramEnv):
```
Не хватает тх.

seriousidea commented on examples/example.py:
```diff
@@ -0,0 +1,73 @@
+    time: datetime.datetime,
+    bids: Sequence,
+    asks: Sequence
+):
```
Не хватает тх.

seriousidea commented on src/galts_trade_api/terminal.py:
```diff
@@ -0,0 +1,93 @@
+            self._on_init_exchange_entities_response
+        )
+
+    async def _on_init_exchange_entities_response(
```
Приватные методы должны идти после всех публичных?

seriousidea commented on src/galts_trade_api/transport/__init__.py:
```diff
@@ -0,0 +1,111 @@
+                raise message[0]
+
+            if len(message) != 2:
+                raise ValueError('Pipe response message can contain exactly 2 elements')
```
Лучше не "может", а "должен"? То есть `can` -> `should`.

nevmerzhitsky replied:
Но в случае передачи исключения используется список из одного элемента. Что думаешь?

И, кстати, для передачи исключения можно вообще не использовать список, а чисто объект исключения передавать. Что думаешь?

seriousidea replied:
> Но в случае передачи исключения используется список из одного элемента

Но эта проверка на количество элементов происходит после того, как стало точно известно что это сообщение не содержит исключение. А раз оно не содержит исключение, значит количество элементов в сообщении должно (should) быть 2.

> для передачи исключения можно вообще не использовать список

Не знаю какой вариант лучше, кажется что оба имеют как преимущества, так и недостатки, а в среднем паритет.

nevmerzhitsky committed:
Remove unused imports

nevmerzhitsky committed:
Use set instead of tuple

nevmerzhitsky committed:
Don't handle SIGHUP by default

The signal is used to reload the program in modern times.

nevmerzhitsky committed:
Fix arg name and add type-hint

nevmerzhitsky committed:
Add type-hints

nevmerzhitsky committed:
Add type-hints

nevmerzhitsky committed:
Convert signal_handler() from coro to simple function

nevmerzhitsky committed:
Make shutdown() more adult

Copy-paste logic from asyncio.run(): now it will log unhandled
exceptions from cancelled tasks.

nevmerzhitsky committed:
Log a process ID to simplify debug

nevmerzhitsky committed:
Simplify main loop of a transport process by applying the new way code

nevmerzhitsky committed:
Replace custom exception handling logic from the transport process

Use global exception handler of the process like the main process does.

nevmerzhitsky committed:
Add semantic sugar

nevmerzhitsky committed:
Rename a class

nevmerzhitsky committed:
Order public and private methods

nevmerzhitsky committed:
Don't close another's event loop

seriousidea commented on src/galts_trade_api/tools.py:
```diff
@@ -0,0 +1,33 @@
+        return cls._instances[key]
+
+    @classmethod
+    def _calculate_args_hash(mcs, args: Tuple, kwargs: Dict) -> int:
```
Кстати, в PEP рекомендуют для таких "контейнеров" как `args` и `kwargs` указывать не тип этого контейнера, а тип одного значения, хранящегося в нем:
https://stackoverflow.com/a/37032111

nevmerzhitsky replied:
Клевая инфа! Но в отмеченной тобой строчке у аргументов нет `*` или `**` (вместо args и kwargs может любое другое быть написано), это рядовые переменные, так что здесь такое не сработает.

seriousidea replied:
Согласен. Однако, во всех остальных случаях, когда используются символы упаковки, тайп хинты отсутствуют. Значит, предлагаю указывать тайп хинты для `*args` и `**kwargs` опционально (можно указать, но не обязательно).

nevmerzhitsky committed:
Little tune a comment

nevmerzhitsky committed:
Rename a var

nevmerzhitsky committed:
Rename a field

nevmerzhitsky committed:
Rename methods to bring more clear semantic

nevmerzhitsky committed:
Move a private method to the tail of a class

nevmerzhitsky committed:
Rename a method to reduce ambiguity

nevmerzhitsky committed:
Move a private method to the tail of a class

nevmerzhitsky commented:
@seriousidea я все твои комментарии обработал.

nevmerzhitsky committed:
Add ability to set debug mode for the loop in the transport process

nevmerzhitsky committed:
Fix a bug after a field rename

nevmerzhitsky committed:
Fix a bug after a variable rename

nevmerzhitsky committed:
Move subscription methd from Market to Terminal

Terminal class can fill this concern because DepthConsumeKey can filter
depths by any kind: exchange, symbol, or market.

nevmerzhitsky committed:
Don't define special exception handler

Now there is no object which require manual clean-up

nevmerzhitsky committed:
Fix import and use it in the example

nevmerzhitsky committed:
Replace prints by logger

seriousidea commented on src/galts_trade_api/transport/real.py:
```diff
@@ -0,0 +1,286 @@
+
+        return self._handlers[request_type]
+
+    def _response_owner_request(self, request: PipeRequest, content: Any):
```
`response owner request`

Гугл сходу переводит это как "ответ владельца запроса" ( == response of request owner), что кажется очевидно не верным.

Хорошо бы устранить неоднозначность формулировки. 

Из содержимого этого метода понятно, что в названии имеется в виду "ответить владельцу запроса". В таком случае можно было бы устранить двусмысленность, например так:
`response to owner request`
Так будет более однозначно, но не на 100%, т.к. слово response - это существительное, а кажется что тут нужен глагол, то есть лучше так:
`respond to owner request`
А ещё лучше так:
`respond to request owner`
Потому что:

request owner == владелец запроса == owner of request
owner request == запрос владельца == ERROR

keyboard button == кнопка клавиатуры == button of keyboard
button keyboard == кнопочная клавиатура == ERROR

nevmerzhitsky replied:
Про то, что response существительное и лучше его заменить на respond to согласен, а с изменением порядка последующих слов - нет. Не понял твои последние примеры перевода: что значит ERROR?

Семантика функции в ответе на запроса процесса-владельца, поэтому это request of owner, а не owner of request. Поэтому owner request (общий подход перемещения слов при удалении "of" из фразы).

seriousidea commented on src/galts_trade_api/asyncio_helper.py:
```diff
@@ -0,0 +1,118 @@
+
+
+# Do it like asyncio.run()
+async def _cancel_tasks(loop: asyncio.AbstractEventLoop, tasks: List) -> None:
```
Кстати о расположении приватных методов в конце. Это правило касается только методов, но не функций?

nevmerzhitsky replied:
Не знаю, мне кажется относительно функций не стоит такое правило добавлять.

seriousidea commented on src/galts_trade_api/asyncio_helper.py:
```diff
@@ -0,0 +1,118 @@
+) -> None:
+    """
+    Args:
+        - handle_signals: If None then these signals will be handled: SIGTERM, SIGINT.
```
Тире не нужно указывать, судя по гугловским стайлгайдам.

nevmerzhitsky committed:
Fix word in message

nevmerzhitsky committed:
Improve semantic of a function name

nevmerzhitsky committed:
Fix docstring code-style
